dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/loadbalance/ConsistentHashLoadBalance.java

```diff
@@ -32,114 +33,156 @@
-                argumentIndex[i] = Integer.parseInt(index[i]);
-            }
-            for (Invoker<T> invoker : invokers) {
-                String address = invoker.getUrl().getAddress();
-                for (int i = 0; i < replicaNumber / 4; i++) {
-                    byte[] digest = md5(address + i);
-                    for (int h = 0; h < 4; h++) {
-                        long m = hash(digest, h);
-                        virtualInvokers.put(m, invoker);
-                    }
-                }
-            }
-        }
-
-        public Invoker<T> select(Invocation invocation) {
-            String key = toKey(invocation.getArguments());
-            byte[] digest = md5(key);
-            return selectForKey(hash(digest, 0));
-        }
-
-        private String toKey(Object[] args) {
-            StringBuilder buf = new StringBuilder();
-            for (int i : argumentIndex) {
-                if (i >= 0 && i < args.length) {
-                    buf.append(args[i]);
-                }
-            }
-            return buf.toString();
-        }
-
-        private Invoker<T> selectForKey(long hash) {
-            Invoker<T> invoker;
-            Long key = hash;
-            if (!virtualInvokers.containsKey(key)) {
-                SortedMap<Long, Invoker<T>> tailMap = virtualInvokers.tailMap(key);
-                if (tailMap.isEmpty()) {
-                    key = virtualInvokers.firstKey();
-                } else {
-                    key = tailMap.firstKey();
-                }
-            }
-            invoker = virtualInvokers.get(key);
-            return invoker;
-        }
-
-        private long hash(byte[] digest, int number) {
-            return (((long) (digest[3 + number * 4] & 0xFF) << 24)
-                    | ((long) (digest[2 + number * 4] & 0xFF) << 16)
-                    | ((long) (digest[1 + number * 4] & 0xFF) << 8)
-                    | (digest[number * 4] & 0xFF))
-                    & 0xFFFFFFFFL;
-        }
-
-        private byte[] md5(String value) {
-            MessageDigest md5;
-            try {
-                md5 = MessageDigest.getInstance("MD5");
-            } catch (NoSuchAlgorithmException e) {
-                throw new IllegalStateException(e.getMessage(), e);
-            }
-            md5.reset();
-            byte[] bytes;
-            try {
-                bytes = value.getBytes("UTF-8");
-            } catch (UnsupportedEncodingException e) {
-                throw new IllegalStateException(e.getMessage(), e);
-            }
-            md5.update(bytes);
-            return md5.digest();
-        }
-
-    }
+	private final ConcurrentMap<String, ConsistentHashSelector<?>> selectors = new ConcurrentHashMap<String, ConsistentHashSelector<?>>();
+
+
+	@Override
+	@SuppressWarnings("unchecked")
+	protected <T> Invoker<T> doSelect(List<Invoker<T>> invokers, URL url, Invocation invocation) {
+		String methodName = RpcUtils.getMethodName(invocation);

```

Comment:
Please indent all code by 4 spaces

Reply:
I'm so sorry that the code has been indented by 4 space in my IDE...emmm...